File: 2019/day02/java/Solution.java
import java.io.*;
import java.util.*;

public class Solution {

    private static final String INPUT_TXT = "input.txt";

    public static List<Integer> getContent(final String path) {
        List<Integer> program = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
            String line = reader.readLine();
            if (line != null) {
                String[] parts = line.split(",");
                for (String part : parts) {
                    program.add(Integer.parseInt(part));
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return program;
    }

    public static int runProgram(List<Integer> program) {
        int i = 0;
        while (program.get(i) != 99) {
            int opcode = program.get(i);
            int pos1 = program.get(i + 1);
            int pos2 = program.get(i + 2);
            int pos3 = program.get(i + 3);
            if (opcode == 1) {
                program.set(pos3, program.get(pos1) + program.get(pos2));
            } else if (opcode == 2) {
                program.set(pos3, program.get(pos1) * program.get(pos2));
            }
            i += 4;
        }
        return program.get(0);
    }

    public static int partOne(List<Integer> program) {
        List<Integer> programCopy = new ArrayList<>(program);
        programCopy.set(1, 12);
        programCopy.set(2, 2);
        return runProgram(programCopy);
    }

    public static int partTwo(List<Integer> program) {
        for (int noun = 0; noun <= 99; noun++) {
            for (int verb = 0; verb <= 99; verb++) {
                List<Integer> programCopy = new ArrayList<>(program);
                programCopy.set(1, noun);
                programCopy.set(2, verb);
                int result = runProgram(programCopy);
                if (result == 19690720) {
                    return 100 * noun + verb;
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        List<Integer> program = getContent(INPUT_TXT);

        int resultPartOne = partOne(program);
        System.out.println("Part 1: " + resultPartOne);

        int resultPartTwo = partTwo(program);
        System.out.println("Part 2: " + resultPartTwo);
    }
}

File: 2019/day02/javascript/solution.js
// Solution to https://adventofcode.com/2019/day/2
const fs = require("fs");

// Convert input.txt in array
// @ts-ignore: ignore ts rules
function inputToArray(input) {
    console.log(input);
    const inputArr = fs.readFileSync(input).toString().split("\n");
    if (inputArr[inputArr.length - 1] == '') {
        inputArr.pop();
    }
    const outputArr = inputArr[0].split(',').map(Number);
    return outputArr;
}

// @ts-ignore: ignore ts rules
function runProgram(program) {
    let num1, num2, result;
    let num1Pos, num2Pos, resultPos;

    for (let i = 0; i < program.length; i += 4) {
        num1Pos = program[i + 1];
        num2Pos = program[i + 2];
        resultPos = program[i + 3];

        num1 = program[num1Pos];
        num2 = program[num2Pos];

        if (program[i] === 1) {
            result = num1 + num2;
        } else if (program[i] === 2) {
            result = num1 * num2;
        } else if (program[i] === 99) {
            break;
        } else {
            console.log("Error: opcode not found!");
        }
        program[resultPos] = result;
    }
    return program;
}

// @ts-ignore: ignore ts rules
function solveA(inputPath) {
    // push input in array
    const inputArr = inputToArray(inputPath);

    // prepare program
    inputArr[1] = 12;
    inputArr[2] = 2;
    console.log("Part One: " + runProgram(inputArr)[0]);
}

// @ts-ignore: ignore ts rules
function solveB(inputPath) {
    const seek = 19690720;
    let found = false;

    for (let verb = 0; verb < 100; verb++) {
        for (let noun = 0; noun < 100; noun++) {
            // push input in array
            const inputArr = inputToArray(inputPath);

            // prepare program
            inputArr[1] = noun;
            inputArr[2] = verb;

            const program = runProgram(inputArr);
            // console.log(program[0]);
            if (program[0] === seek) {
                found = true;
                console.log("Part Two: Number " + seek + " found (" + found +") with verb " + verb + " and noun " + noun + " (Result: " + (100 * noun + verb) + ").");
            }
        }
    }
}

function runTests() {
    const tests = {
        "TestA1": [
            [1, 0, 0, 0, 99],
            [2, 0, 0, 0, 99]
        ],
        "TestA2": [
            [2, 3, 0, 3, 99],
            [2, 3, 0, 6, 99]
        ],
        "TestA3": [
            [2, 4, 4, 5, 99, 0],
            [2, 4, 4, 5, 99, 9801]
        ],
        "TestA4": [
            [1, 1, 1, 4, 2, 5, 6, 0, 99],
            [30, 1, 1, 4, 2, 5, 6, 0, 99]
        ]
    };
    let succeed = true;

    for (const key in tests) {
        // @ts-ignore: ignore ts rules
        const result = runProgram(tests[key][0]);
        // @ts-ignore: ignore ts rules
        const expResult = tests[key][1];
        if (JSON.stringify(result) != JSON.stringify(expResult)) {
            succeed = false;
        }
    }
    return succeed;
}

if (runTests() === true) {
    solveA('../input.txt');
    solveB('../input.txt');
}
